### PR TITLE
feat: isolate token balance fetch

### PR DESCRIPTION
**Pull request by janek26**
This PR reduces the calls to the sequencer from 4 to 1 for each token to update the balance.
It also removes old clunky patterns and uses zustand instead

### CODE REVIEW

File: packages/extension/src/ui/states/tokens.ts [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+                key,
+                (value as TokenDetails[]).filter(
+                  (token) =>
+                    !parsedDefaultErc20Tokens.some((defaultToken) =>
+                      equalToken(token, defaultToken),
+                    ),
+                ),
+              ]
+            }
+            return [key, value]
```
Comment (delaaxe) [inline packages/extension/src/ui/states/tokens.ts:72]: A bit spaghetti-ish

File: packages/extension/src/ui/states/tokens.ts [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+
+export const addToken = (token: TokenDetails) => {
+  useTokens.getState().addToken(token)
+}
+
+export const selectorTokensByNetwork =
+  (networkId: string) => (state: TokenState) =>
+    state.tokens.filter((token) => token.networkId === networkId)
+
+export interface TokenDetailsWithBalance extends TokenDetails {
```
Comment (delaaxe) [inline packages/extension/src/ui/states/tokens.ts:87]: nice

File: packages/extension/src/ui/states/tokens.ts [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+        ),
+    },
+  ),
+)
+
+export const addToken = (token: TokenDetails) => {
+  useTokens.getState().addToken(token)
+}
+
+export const selectorTokensByNetwork =
```
Comment (delaaxe) [inline packages/extension/src/ui/states/tokens.ts:83]:
```suggestion
export const selectTokensByNetwork =
```

File: packages/extension/src/ui/states/tokens.ts [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+
+  const { data, isValidating, error, mutate } = useSWR(
+    [walletAddress, ...tokenAddresses],
+    async (walletAddress, ...tokenAddresses) => {
+      const balances = await Promise.all(
+        tokenAddresses.map(async (address) =>
+          fetchTokenBalance(address, walletAddress, networkId),
+        ),
+      )
+      return balances.reduce((acc, balance, i) => {
```
Comment (delaaxe) [inline packages/extension/src/ui/states/tokens.ts:109]: Avoid using `reduce`...?

File: packages/extension/src/ui/utils/tokens.ts [hunk 5]
```diff
@@ -1,80 +1,17 @@
...
-
-export interface TokenDetails {
-  address: string
-  name?: string
-  symbol?: string
-  decimals?: BigNumber
-  balance?: BigNumber
-  networkId: string
-}
-
```
Comment (delaaxe) [inline packages/extension/src/ui/utils/tokens.ts:77]: good riddance, that whole part was a little clunky

File: packages/extension/src/ui/utils/tokens.ts [hunk 6]
```diff
@@ -131,19 +68,29 @@ export const fetchTokenDetails = async (
...
+): Promise<BigNumber> => {
+  const provider = getProvider(networkId)
+  const tokenContract = new Contract(parsedErc20Abi as Abi[], address, provider)
+  const balance = await tokenContract
+    .call("balanceOf", { user: walletAddress })
+    .then((x) =>
+      BigNumber.from(uint256.uint256ToBN(x.balance as any).toString()),
+    )
+    .catch(() => BigNumber.from(0))
+  return balance
```
Comment (delaaxe) [inline packages/extension/src/ui/utils/tokens.ts:95]: Callback based programming... no

```suggestion
  try {
    const result = await tokenContract.call("balanceOf", { user: walletAddress })
    return BigNumber.from(uint256.uint256ToBN(result.balance as any).toString())
  } catch {
    return BigNumber.from(0)
  }
```

**Review by delaaxe — APPROVED** [review]

File: packages/extension/src/ui/states/tokens.ts (hunk 1 above)
Comment (janek26) [reply packages/extension/src/ui/states/tokens.ts:72]: agree, but I wanted to avoid introducing `immer` as a dependency

**Pull request merged**